core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala

```diff
@@ -114,28 +123,45 @@ class HistoryServer(
    * this UI with the event logs in the provided base directory.
    */
   def initialize() {
-    attachPage(new HistoryPage(this))
+    if (!initialized.getAndSet(true)) {
```

Comment:
If that's the case, `initialize()` should be made private and its calls removed from the tests.

(I've seen this pattern of calling an `init()` method from constructors in Scala when the initialization code requires local variables, which otherwise would have to be fields in the class...)